src/MessagePack/Formatters/TypelessFormatter.cs

```diff
@@ -0,0 +1,293 @@
+
+        public static readonly IMessagePackFormatter<object> Instance = new TypelessFormatter();
+
+        static readonly Dictionary<Type, KeyValuePair<object, SerializeMethod>> serializers = new Dictionary<Type, KeyValuePair<object, SerializeMethod>>();
+        static readonly Dictionary<Type, KeyValuePair<object, DeserializeMethod>> deserializers = new Dictionary<Type, KeyValuePair<object, DeserializeMethod>>();
+
+        readonly HashSet<string> blacklistCheck;
+
+        /// <summary>
+        /// When type name does not have Version, Culture, Public token - sometimes can not find type, example - ExpandoObject
+        /// In that can set to `false`
+        /// </summary>
+        public static volatile bool RemoveAssemblyVersion = true;
+
+        TypelessFormatter()
+        {
+            blacklistCheck = new HashSet<string>()
+            {
+                "System.CodeDom.Compiler.TempFileCollection",
+                "System.IO.FileSystemInfo",
+                "System.Management.IWbemClassObjectFreeThreaded"
+            };
+        }
+
+        // see:http://msdn.microsoft.com/en-us/library/w3f99sx1.aspx
+        // subtract Version, Culture and PublicKeyToken from AssemblyQualifiedName 
+        internal static string BuildTypeName(Type type)
+        {
+            if (RemoveAssemblyVersion)
+                return SubtractFullNameRegex.Replace(type.AssemblyQualifiedName, "");
+            else
+                return type.AssemblyQualifiedName;
+        }
+
+        public int Serialize(ref byte[] bytes, int offset, object value, IFormatterResolver formatterResolver)
+        {
+            if (value == null)
+            {
+                return MessagePackBinary.WriteNil(ref bytes, offset);
+            }
+
+            if (value is IDictionary) // check IDictionary first
+            {
+                var startOffset = offset;
+                var d = value as IDictionary;
+                offset += MessagePackBinary.WriteMapHeader(ref bytes, offset, d.Count);
+                foreach (DictionaryEntry item in d)
+                {
+                    offset += SerializeItem(ref bytes, offset, item.Key, formatterResolver);
+                    offset += SerializeItem(ref bytes, offset, item.Value, formatterResolver);
+                }
+                return offset - startOffset;
+            }
+            else if (value is ICollection)
+            {
+                var startOffset = offset;
+                var c = value as ICollection;
+                offset += MessagePackBinary.WriteArrayHeader(ref bytes, offset, c.Count);
+                foreach (var item in c)
+                {
+                    offset += SerializeItem(ref bytes, offset, item, formatterResolver);
+                }
+                return offset - startOffset;
+            }
+            else
+            {
+                return SerializeItem(ref bytes, offset, value, formatterResolver);
```

Comment:
Where is the PrimitiveObjectResolver check?
Note: PrimitiveObjectResolver.IsSupportedType(DateTime) returns true, but it losts Kind info so check DateTime is good.